[WIP] Add Wolf-Waagen API

tillprochaska opened the pull request.

tillprochaska committed:
Add a wrapper around the Wolf Waagen API

makes it easy to send a request and to normalize data

tillprochaska committed:
Make some params optional

tillprochaska committed:
Add hive id to report settings

tillprochaska committed:
Add rake task to set up sensor types

roschaefer committed:
Beautiful implementation for wolf_waagen_api:fetch

roschaefer committed:
Fix rubocop

tillprochaska committed:
Add rake task to import latest sensor readings

roschaefer commented on app/models/report.rb:
```diff
@@ -11,6 +11,8 @@ class Report < ActiveRecord::Base
   has_many :variables, dependent: :destroy
   accepts_nested_attributes_for :variables
 
+  validates :hive_id, presence: true
```
```suggestion
```

roschaefer commented on app/lib/wolf_waagen_api/request.rb:
```diff
@@ -0,0 +1,54 @@
+    def initialize(hive_id:, start_date:, end_date:, interval: 'hour')
+      @hive_id = hive_id
+
+      raise ArgumentError, '`start_date` and `end_date` should be instances of `DateTime`.' unless [start_date, end_date].all? DateTime
```
I believe you have some overly strict validations here. My suggestion is to write a test first, reproduce the RuntimeError and *then* think about a good solution. E.g. validate the class *or* even better you transform the given object.

roschaefer commented on app/lib/wolf_waagen_api/result.rb:
```diff
@@ -0,0 +1,32 @@
+
+    def series
+      @raw_series
+        .select { |series| series[:values].is_a? Array }
```
In what occasions `@raw_series` is not an Array?

tillprochaska replied:
If there’s not data available, the API returns `nil`.

roschaefer commented on app/lib/wolf_waagen_api/series.rb:
```diff
@@ -0,0 +1,36 @@
+
+    def points
+      points = @values.map.with_index do |value, index|
+        Point.new(series: self, index: index, data: value)
```
*Every* point must get a reference to the `series` :thinking:

roschaefer commented on spec/lib/wolf_waagen_api/series.rb:
```diff
@@ -0,0 +1,91 @@
+      end.to raise_error ArgumentError
+    end
+
+    it 'should allow unitless series' do
```
```suggestion
    it 'allows unit-less series' do
```

roschaefer commented on spec/lib/wolf_waagen_api/series.rb:
```diff
@@ -0,0 +1,91 @@
+      end.not_to raise_error
+    end
+
+    it 'should allow series without accuracy' do
```
```suggestion
    it 'allows series without accuracy' do
```

roschaefer reviewed (changes requested):
Many many restrictions and validation. Try to be more lenient next time.

roschaefer committed:
Permit :hive_id

roschaefer committed:
Fix rubocop

roschaefer committed:
Validations in report will break old migrations

Those migrations create static data and do not anticipate any future
validations.

roschaefer committed:
Merge remote-tracking branch 'origin/master' into 2-add-wolf-waagen-api

roschaefer committed:
Fix rubocop

roschaefer committed:
Fix rspec
